File: tests/test_api.py
from unittest import TestCase
from vsi.api import lex, parse, interpret
from vsi.lexer import Token
from vsi.ast import AST, AssignStatement, Integer


class TestApi(TestCase):

    def test_lex_function(self):
        string = "x := 10;"
        tokens = lex(string)
        expected_tokens = [Token("x", "ID", None, 0),
                           Token(":=", "RESERVED", "OPERATOR", 0),
                           Token(10, "NUMBER", "INT", 0),
                           Token(";", "RESERVED", "SYMBOL", 0),
                           Token("EOF", "RESERVED", "SYMBOL", 0)
                           ]

        self.assertEqual(tokens, expected_tokens)

    def test_parse_function(self):
        string = "x := 10;"
        ast = parse(string)
        expected_ast = AST([AssignStatement("x", Integer(10))])
        self.assertEqual(ast, expected_ast)

    def test_interpret_function(self):
        string = "x := 10;"
        env = {}
        interpret(string, env)
        self.assertIn("x", env)
        self.assertEqual(env["x"], 10)

File: Vsi.Tests/TestApi.cs
using System.Collections.Generic;
using Xunit;
using Vsi.Core;

namespace Vsi.Tests
{
    public class TestApi
    {
        [Fact]
        public void TestLexFunction()
        {
            string str = "x := 10;";
            var tokens = Api.Lex(str);
            var expectedTokens = new List<Token>
            {
                new Token("x", TokenType.Id, TokenSpecificType.None, 0),
                new Token(":=", TokenType.Reserved, TokenSpecificType.Operator, 0),
                new Token(10, TokenType.Number, TokenSpecificType.Int, 0),
                new Token(";", TokenType.Reserved, TokenSpecificType.Symbol, 0),
                new Token("EOF", TokenType.Reserved, TokenSpecificType.Symbol, 0)
            };

            Assert.Equal(expectedTokens, tokens);
        }

        [Fact]
        public void TestParseFunction()
        {
            string str = "x := 10;";
            var ast = Api.Parse(str);
            var expectedAst = new Ast(new List<AstNode> { new AssignStatement("x", new Integer(10)) });
            Assert.Equal(expectedAst, ast);
        }

        [Fact]
        public void TestInterpretFunction()
        {
            string str = "x := 10;";
            var env = new Dictionary<string, object?>();
            Api.Interpret(str, env);
            Assert.True(env.ContainsKey("x"));
            Assert.Equal(10, env["x"]);
        }
    }
}
